packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart

```diff
@@ -40,17 +40,79 @@ class PlatformCameraPosition {
 
 /// Pigeon representation of a CameraUpdate.
 class PlatformCameraUpdate {
-  PlatformCameraUpdate(this.json);
-
-  /// The update data, as JSON. This should only be set from
-  /// CameraUpdate.toJson, and the native code must interpret it according to the
-  /// internal implementation details of the CameraUpdate class.
-  // TODO(stuartmorgan): Update the google_maps_platform_interface CameraUpdate
-  //  class to provide a structured representation of an update. Currently it
-  //  uses JSON as its only state, so there is no way to preserve structure.
-  //  This wrapper class exists as a placeholder for now to at least provide
-  //  type safety in the top-level call's arguments.
-  final Object json;
+  PlatformCameraUpdate({
+    this.newCameraPosition,
+    this.newLatLng,
+    this.newLatLngZoom,
+    this.newLatLngBounds,
+    this.scrollBy,
+    this.zoomBy,
+    this.zoomIn = false,
+    this.zoomOut = false,
+    this.zoomTo,
+  });
+
+  final PlatformNewCameraPosition? newCameraPosition;
+  final PlatformNewLatLng? newLatLng;
+  final PlatformNewLatLngBounds? newLatLngBounds;
+  final PlatformNewLatLngZoom? newLatLngZoom;
+  final PlatformScrollBy? scrollBy;
+  final PlatformZoomBy? zoomBy;
+  final bool zoomIn;
+  final bool zoomOut;
+  final PlatformZoomTo? zoomTo;
```

Comment:
This isn't the design we discussed in the issue; what's the reason for the change?

Reply:
If the change you are referring to is `PlatformCameraUpdate` containing members instead of each option extending it as a base class, this is due to the limitations of Pigeon, as we discussed privately. If each of the individual subclasses extend `PlatformCameraUpdate`, but are not reachable by the exposed APIs, they will not be converted by Pigeon at all, and therefore be unusable. The only other alternative I can think of that still keeps the subtypes reachable is to overload `animate/moveCamera` for each of the subtypes, which does not sound very preferable.

Reply:
Oh, I didn't realize that was the context for that question.

@tarrinneal How hard would it be to generate all classes in the current implementation? I never remember if it works this way because it used to have to when Pigeon used mirrors, and it just stayed that way, or if it had to change to this way when we moved *off* of mirrors.

Reply:
I'm not sure I'm understanding the question here. @stuartmorgan

Reply:
@tarrinneal I believe he is basically asking the following. Suppose there is some pigeon defined class `X` which is exposed via APIs and is therefore converted by the code generation when running pigeon. Is there a way to define multiple subclasses of `X` that will also be converted by code gen without defining new functions that explicitly use them?

Reply:
We chatted about this briefly in triage yesterday, and @tarrinneal believes this would be easy to fix in Pigeon, in which case we should probably do that and then update this PR accordingly.

Reply:
Okay, so how would we do that? By "fix in Pigeon", do you mean pushing a change to Pigeon itself?

Reply:
I'll push the pigeon fix asap

Reply:
https://github.com/flutter/packages/pull/7529

Reply:
@stuartmorgan It sounds like what Tarrin has fixed in Pigeon does not actually resolve this, as inheritance is not supported in Pigeon in any capacity.

Reply:
Adding inheritance is a much bigger project I don't think that is something that can ready in a reasonable time frame for this pr.

Reply:
Could we or do we want to bypass inheritance and say that the cameraUpdate object contains an Object and an a defined enum for the type. Then in methods that take cameraUpdate switch on an enum and cast.

Reply:
> I think I am still blocked. It looks like Pigeon's generated code does not preserve inheritance between classes

This is why the issue describes using `Object` as the item type. Inheritance would be nice, but shouldn't be blocking.

> Could we or do we want to bypass inheritance and say that the cameraUpdate object contains an Object and an a defined enum for the type. Then in methods that take cameraUpdate switch on an enum and cast.

We don't even need the enum; we can still do type detection on the class without inheritance.

We could do it if you feel strongly, but I think it's just adding duplicate information. Checking the actual type before casting to that type seems much better to me than checking an enum and then casting.